data/transactions/currency.go

```diff
@@ -0,0 +1,360 @@
+		// Changing keys in a currency.
+		if !params.Manager.IsZero() {
+			params.Manager = cc.CurrencyParams.Manager
+		}
+		if !params.Reserve.IsZero() {
+			params.Reserve = cc.CurrencyParams.Reserve
+		}
+		if !params.Freeze.IsZero() {
+			params.Freeze = cc.CurrencyParams.Freeze
+		}
+		if !params.Clawback.IsZero() {
+			params.Clawback = cc.CurrencyParams.Clawback
+		}
+
+		record.CurrencyParams[cc.ConfigCurrency.Index] = params
+	}
+
+	return balances.Put(record)
+}
+
+func (ct CurrencyTransferTxnFields) apply(header Header, balances Balances, spec SpecialAddresses, ad *ApplyData) error {
+	// Default to sending from the transaction sender's account.
+	source := header.Sender
+	clawback := false
+
+	if !ct.CurrencySender.IsZero() {
+		// Clawback transaction.  Check that the transaction sender
+		// is the Clawback address for this currency.
+		params, err := getParams(balances, ct.XferCurrency)
+		if err != nil {
+			return err
+		}
+
+		if header.Sender != params.Clawback {
+			return fmt.Errorf("clawback not allowed: sender %v != clawback %v", header.Sender, params.Clawback)
+		}
+
+		// Transaction sent from the correct clawback address,
+		// execute currency transfer from specified source.
+		source = ct.CurrencySender
+		clawback = true
+	}
+
+	if ct.CurrencyAmount > 0 || ct.CurrencyReceiver != (basics.Address{}) {
+		snd, err := balances.Get(source)
+		if err != nil {
+			return err
+		}
+
+		snd.Currencies = clone(snd.Currencies)
+		sndHolding, ok := snd.Currencies[ct.XferCurrency]
+		if !ok {
+			// Allocating a slot for currency (self-transfer of zero amount).
+			// Initialize holding with default Frozen value.
+			if clawback {
+				return fmt.Errorf("cannot allocate currency slot via clawback")
+			}
+
+			params, err := getParams(balances, ct.XferCurrency)
+			if err != nil {
+				return err
+			}
+
+			sndHolding.Frozen = params.DefaultFrozen
+		}
+
+		if sndHolding.Amount < ct.CurrencyAmount {
+			return fmt.Errorf("currency balance %d less than transfer amount %d", sndHolding.Amount, ct.CurrencyAmount)
+		}
+
+		if ct.CurrencyAmount > 0 && sndHolding.Frozen && !clawback {
+			return fmt.Errorf("currency frozen in sender")
+		}
+
+		sndHolding.Amount -= ct.CurrencyAmount
+		snd.Currencies[ct.XferCurrency] = sndHolding
+		err = balances.Put(snd)
+		if err != nil {
+			return err
+		}
+
+		if len(snd.Currencies) > balances.ConsensusParams().MaxCurrenciesPerAccount {
+			return fmt.Errorf("too many currencies in account: %d > %d", len(snd.Currencies), balances.ConsensusParams().MaxCurrenciesPerAccount)
+		}
+
+		rcv, err := balances.Get(ct.CurrencyReceiver)
+		if err != nil {
+			return err
+		}
+
+		rcv.Currencies = clone(rcv.Currencies)
+		rcvHolding, ok := rcv.Currencies[ct.XferCurrency]
+		if !ok {
+			return fmt.Errorf("currency not present in receiver account")
+		}
+
+		if ct.CurrencyAmount > 0 && rcvHolding.Frozen {
+			return fmt.Errorf("currency frozen in recipient")
+		}
+
+		rcvHolding.Amount += ct.CurrencyAmount
```

Comment:
there's no risk of overflow here, but might be good to have OAdd as a sanity check

Reply:
Sure, done.